src/nbl/asset/interchange/CGLTFLoader.cpp

```diff
@@ -0,0 +1,1638 @@
+
+									regionOffsets += fullMipMapExtent.x * fullMipMapExtent.y * fullMipMapExtent.z * newImageParams.arrayLayers * asset::getTexelOrBlockBytesize(newImageParams.format);
+								}
+								newCpuBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(regionOffsets);
+
+								newDerivativeNormalMapImage = ICPUImage::create(std::move(newImageParams));
+								newDerivativeNormalMapImage->setBufferAndRegions(std::move(newCpuBuffer), newRegions);
+
+								using DerivativeNormalMapFilter = CNormalMapToDerivativeFilter;
+								DerivativeNormalMapFilter derivativeNormalFilter;
+								DerivativeNormalMapFilter::state_type state;
+
+								state.inImage = cpuImageNormalTexture.get();
+								state.outImage = newDerivativeNormalMapImage.get();
+								state.inOffset = { 0, 0, 0 };
+								state.inBaseLayer = 0;
+								state.outOffset = { 0, 0, 0 };
+								state.outBaseLayer = 0;
+								state.extent = { referenceImageParams.extent.width, referenceImageParams.extent.height, referenceImageParams.extent.depth };
+								state.layerCount = newDerivativeNormalMapImage->getCreationParameters().arrayLayers;
+
+								state.scratchMemoryByteSize = state.getRequiredScratchByteSize(state.extent);
+								state.scratchMemory = reinterpret_cast<uint8_t*>(_NBL_ALIGNED_MALLOC(state.scratchMemoryByteSize, 32));
+
+								state.inMipLevel = 0;
+								state.outMipLevel = 0;
+
+								if (!derivativeNormalFilter.execute(&state))
+									os::Printer::log("Something went wrong while performing derivative filter operations!", ELL_ERROR);
+
+								_NBL_ALIGNED_FREE(state.scratchMemory);
+
+								ICPUImageView::SCreationParams imageViewInfo;
+								imageViewInfo.image = std::move(newDerivativeNormalMapImage);
+								imageViewInfo.format = imageViewInfo.image->getCreationParameters().format;
+								imageViewInfo.viewType = decltype(imageViewInfo.viewType)::ET_2D;
+								imageViewInfo.flags = static_cast<ICPUImageView::E_CREATE_FLAGS>(0u);
+								imageViewInfo.subresourceRange.baseArrayLayer = 0u;
+								imageViewInfo.subresourceRange.baseMipLevel = 0u;
+								imageViewInfo.subresourceRange.layerCount = newDerivativeNormalMapImage->getCreationParameters().arrayLayers;
+								imageViewInfo.subresourceRange.levelCount = newDerivativeNormalMapImage->getCreationParameters().mipLevels;
+
+								auto imageView = ICPUImageView::create(std::move(imageViewInfo));
+
+								if(!imageView.get())
+									os::Printer::log("Something went wrong while creating image view for derivative normal map!", ELL_ERROR);
+
+								/*
+									@devshgraphicsprogramming
+
+									TODO: I think it should be done at the very begining of the glTF loading resources since it may occur
+									for a single normal map will be applied more than one derivative computations.
+
+									Such asset should also store info about `scale` from state for derivatives in a geometry shader
```

Comment:
and especially this one @Devsh-Graphics-Programming/seniors

Reply:
> TODO: I think it should be done at the very begining of the glTF loading resources since it may occur for a single normal map will be applied more than one derivative computations.

This is why every one of @Crisspl's loaders caches the result of any transformations under a cache key which is similar to `your/asset/currently/beingloaded.glTF?derivativeMap?someUniqueTextureID`

He first looks up the cache, and loads the texture + creates the derivative map if it doesn't exist. Then it doesnt matter if you create all the deriv maps at the beginning, end or middle, you're able to catch all the duplicate references.

> Such asset should also store info about `scale` from state for derivatives in a geometry shader

What geometry shader!?

The `scale` for a cached asset... this is why metadata exists!

You should probably solve by creating a "pseudo derivative map loader" with its own metadata type, not to be used just to have a unique string for the metadata type method and a metadata type when you add derivative maps to the cache as their own `SAssetBundle`.

Reply:
Hmm, I wonder if we could wrap this code from here https://github.com/Devsh-Graphics-Programming/Nabla/pull/9#discussion_r599337479 to the derivative map loader and taking normal map as an input (but we could provide a parameter in loading parameters determining whether asset being loaded is normal map or derivative map) and add new second parameter determining which loader uses this loader to provide caching as following:

std::string_view _loader_ = ...;

`your/asset/currently/beingloaded._loader_ ?derivativeMap?someUniqueTextureID`

In my opinion it would clean everything for the future usage, what do you think @Devsh-Graphics-Programming/seniors ?

Reply:
i dont get the idea at all, just cache the result of derivative map creation. I dont understand what would you need those extra loading parameters for

Reply:
regarding this dummy derivative map loader, matt would have to say more

Reply:
because this cache key assumes glTF only and I want to make a global derivative loader

Reply:
with metadata above and deriv filter execution

Reply:
just asking

Reply:
you could make a utility (or just add to already existing derivative map creator), i cant imagine format-agnostic loader. This dummy loader matt was talking about is just to be able to have another metadata type if i understand correctly the idea

Reply:
Its an absolutely awful idea.

An asset loader is a class that sepecializes in loading a part of an asset DAG from one specific file format.

The file format is the only thing that tells the asset loader what the asset is, file extension is more of a hint to try loaders in an appropriate order.

Normal maps come as dds, jpg, png, ktx, all sorts of images, these formats have **absolutely zero information** that the texture you're loading has an intended use as a normal map (a material loader, i.e. an asset loader that loads a ICPUGraphicsPipeline has this sort of info).

Furthermore note that derivative maps are being created "on demand" from an original image w.r.t. a specific asset so there are only two options for the created derivative map's asset cache key:
 - `pipeline_asset_key?derivativeMap?someUUID`
 - `original_image_key?derivativeMap` (better less duplicates, can reuse the reference across loaders, so you shall achieve globalness like you want)

The only thing you can do to reduce code duplication is to add a `IRenderpassIndependentPipelineLoaderBase` class with the methods and utilities to fetch derivative maps from bump or normal maps, such a utility method would:
1) check for the image cache key with `?derivativeMapFromBump` or `?derivativeMapFromNormal` added to the key
2) if (1) not found, then load the source image
3) obtain the real cache key of the loaded source image and try (1) again but this time use a different prefix before the `?`
4) if (3) didnt work then make sure that the source image has been "restored from dummy"
5) create the derivative map with a filter from the source image
5) insert the derivative map into the cache with the same key you tried at (3)

> This dummy loader matt was talking about is just to be able to have another metadata type if i understand correctly the idea

The whole point of this

> You should probably solve by creating a "pseudo derivative map loader" with its own metadata type, **not to be used** just to have a unique string for the metadata type method and a metadata type when you add derivative maps to the cache as their own SAssetBundle.

was that a newly created synthetic result of the deriv map filter with cache key `original_image_key?derivativeMap` so far, in @Crisspl 's usage, has always been inserted into the cache with 0 metadata, nothing.

The idea is to store the `scale` output of the deriv map filter ( the inverse normalization constant)  in some metadata, but the way metadata works is that you need a loader-class-name to compare the metadata when you cast.

So we create a class, that **is not a loader, never loads, never does anything** so that we may create a metadata type which reports some meaningful loader name.